GenXIntrinsics/lib/GenXIntrinsics/GenXSPIRVWriterAdaptor.cpp

```diff
@@ -253,19 +254,19 @@ static Function *
 transformKernelSignature(Function &F, const std::vector<SPIRVArgDesc> &Descs) {
   SmallVector<Type *, 8> NewParams;
 
-  auto GetArgType =
-#if VC_INTR_LLVM_VERSION_MAJOR >= 16
-      [UseTargetTypes = !F.getContext().supportsTypedPointers()](
-          SPIRVArgDesc Desc, Argument &Arg) {
-        if (UseTargetTypes)
-          return getArgTargetTypeFromDesc(Desc, Arg);
-        return getArgTypeFromDesc(Desc, Arg);
-      };
-#else  // VC_INTR_LLVM_VERSION_MAJOR >= 16
-      [](SPIRVArgDesc Desc, Argument &Arg) {
-        return getArgTypeFromDesc(Desc, Arg);
-      };
-#endif // VC_INTR_LLVM_VERSION_MAJOR >= 16
+  bool UseTargetTypes = true;
+  // Before LLVM 16, we don't want to use target types. After LLVM 16, typed
+  // pointers are always disabled, so we must use target types.
+#if VC_INTR_LLVM_VERSION_MAJOR < 16
+  bool UseTargetTypes = false;
+#elif VC_INTR_LLVM_VERSION_MAJOR == 16
+  UseTargetTypes = !F.getContext().supportsTypedPointers();
+#endif
+  auto GetArgType = [UseTargetTypes](SPIRVArgDesc Desc, Argument &Arg) {
+    if (UseTargetTypes)
+      return getArgTargetTypeFromDesc(Desc, Arg);
+    return getArgTypeFromDesc(Desc, Arg);
```

Comment:
I join in the thanks! Thank you very much for a job well done.
One remark - current changes will not build on old llvm (15 and lesser) - I suggest to apply the next patch:
```
@@ -263,17 +263,18 @@ static Function *
 transformKernelSignature(Function &F, const std::vector<SPIRVArgDesc> &Descs) {
   SmallVector<Type *, 8> NewParams;

-  bool UseTargetTypes = true;
   // Before LLVM 16, we don't want to use target types. After LLVM 16, typed
   // pointers are always disabled, so we must use target types.
-#if VC_INTR_LLVM_VERSION_MAJOR < 16
-  bool UseTargetTypes = false;
-#elif VC_INTR_LLVM_VERSION_MAJOR == 16
-  UseTargetTypes = !F.getContext().supportsTypedPointers();
+#if VC_INTR_LLVM_VERSION_MAJOR == 16
+  bool UseTargetTypes = !F.getContext().supportsTypedPointers();
 #endif
-  auto GetArgType = [UseTargetTypes](SPIRVArgDesc Desc, Argument &Arg) {
+  auto GetArgType = [&](SPIRVArgDesc Desc, Argument &Arg) {
+#if VC_INTR_LLVM_VERSION_MAJOR == 16
     if (UseTargetTypes)
       return getArgTargetTypeFromDesc(Desc, Arg);
+#elif VC_INTR_LLVM_VERSION_MAJOR > 16
+    return getArgTargetTypeFromDesc(Desc, Arg);
+#endif
     return getArgTypeFromDesc(Desc, Arg);
   };
```
If you also change comments for lines
```
-#if VC_INTR_LLVM_VERSION_MAJOR < 18
+#if VC_INTR_LLVM_VERSION_MAJOR < 17
...
#endif // VC_INTR_LLVM_VERSION_MAJOR < 18
```
It wold be perfect)

Reply:
Thanks for the review, @igorban-intel.

Ach of course, sorry about the mistake - I hadn't noticed that `getArgTargetTypeFromDesc` was only declared for 16+.

I've made and pushed a few changes, but altered slightly. I was afraid of having `return ... return` producing a "unreachable code" warning on 16+. I have also made `getArgTypeFromDesc` and all of its helper functions compile only for 16- to avoid "unused function" warnings on 17+.

Forgive the simple question, but are you running testing on all of the various LLVM versions to ensure I'm not breaking anything? I haven't actually built or tested with LLVM 15 or 16 so I'm just going by eye right now!

Oh and I updated some of the `#if`/`#endif` comments - sorry for forgetting about them. I didn't know what to do with `#if/#elif/#else/#endif` blocks and couldn't find an example, so left those blank. Do the `#else` and/or `#endif` comments match the most recent `#if`/`#elif`, or the first/outermost one?